### DISCUSSION TITLE
Display SH1106 Freeze

### DISCUSSION THREAD

**Discussion by Snusme83** [hoylabs/OpenDTU-OnBattery#1202] [General]
Servus Community,

habe es heute wieder gewagt, mein OLED SH1106 an den ESP32-S3 zu stecken.

Als erstes habe ich es versucht, einen frischen ESP32-S3 mit der neuen Firmware (2024.08.18) zu bespielen. Wenn nichts an den Pins des Mikrocontrollers hängt, funktioniert das einwandfrei (hatte beim letzten Firmwareupdate Probleme).

An den frisch geupdateten Mikrocontroller habe ich die config.json des aktuell in Betrieb befindlichen, funktionierenden Mikrokontroller aufgespielt. Dann habe ich meine pinmapping. json auf den Mikrocontroller aufgespielt und siehe da, das Display hat mit dieser Konfiguration funktioniert.
[pin_mapping.json](https://github.com/user-attachments/files/16750505/pin_mapping.json)

Da ich das Display auf dem frischen Mikrocontroller in Betrieb nehmen konnte und mir der Netzbezug angezeigt wurde, dachte ich ich spiele die pinmapping.json auf den in Betrieb befindlichen Mikrocontroller auf. Dies funktionierte auch noch erfolgreich.

Als ich das Profil unter den Hardware- Einstellungen ausgewählt hab und anschließend gespeichert hab, hat sich die OpenDTU-OnBattery allerdings wieder eingefroren.

Mit diesem pinmapping.json (ohne Display) funktioniert es einwandfrei:
[pin_mapping.json](https://github.com/user-attachments/files/16750565/pin_mapping.json)

Zu meiner Hardware: 

ESP32-S3 DevKitC N8R2 WROOM1, 2 x Pylontech US3000C, 3 x Victron SmartSolar 100/20, 6 x 370 Wp PV-Module, 1 x Hoymiles HM-1500 Wechselrichter

**Comment by DerHavey1994** [comment]
Ich habe hier das Problem in #1164 schon als Issue gepostet. Ich bin tatsächlich nicht auf die Idee gekommen 3x MPPT & Pylontech ohne Display zu testen. So funktioniert es jetzt tatsächlich ohne Display mit geänderten PIN´s auch bei mir.

@schlimmchen: Es besteht also ein Problem mit zusätzlich angeschlossenem Display. Ist das vielleicht dadurch einfacher das Problem einzugrenzen?
Kann man #1164 & #1202 zusammenführen?

**Reply by Snusme83** [reply]
@DerHavey1994 : 3 x MPPT und Pylontechs zusammen funktioniert bei mir einwandfrei mit der o. g. pinmapping.json!

Erst wenn ich das pinmapping erweitere auf ein Display, dieses aufspiele und das Profil auswähle, friert die DTU ein.

Unbedingt brauche ich das Display ja nicht, aber ich hab es halt zuhause rumliegen und habe ein passendes Gehäuse dafür drucken lassen und in diesem Gehäuse wäre halt alles verstaubar und würde schon schick aussehen!😎

**Reply by DerHavey1994** [reply]
zB 2x MPPT, Pylontech & Display geht auch. Es können wohl nur 4 Geräte gleichzeitig benutzt werden.

**Reply by Snusme83** [reply]
Könnte durchaus sein. Da bin ich gespannt, was die anderen Profis dazu meinen.😉

**Reply by schlimmchen** [reply]
> Kann man https://github.com/helgeerbe/OpenDTU-OnBattery/issues/1164 & https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/1202 zusammenführen?

Das geht nicht weil das eine ein Issue ist und das andere eine Diskussion, aber ja, hier liegt vermutlich das gleiche Ursache vor. #1164 wird auf jeden fall noch angeschaut und sicherlich auch gelöst.

**Reply by Snusme83** [reply]
Das hört sich immer so beruhigend an, wenn schlimmchen das schreibt.

@ schlimmchen: das Flashen der Firmware über webinstall ist einwandfrei möglich, wenn die Pins vom Pylontech nicht auf den Mikrocontroller gesteckt sind. Hatte da immer wieder Probleme!

**Reply by schlimmchen** [reply]
Hast du [UART0 in Gebrauch](https://opendtu-onbattery.net/hardware/limitations/#using-uart0)?

**Reply by Snusme83** [reply]
Wie ist die Frage gemeint? Ist mit UART0 der eine von zwei USB-C Anschlüssen gemeint? Zum Betreiben meiner OpenDTUon Battery ist das glaub ich egal, wo ich anstecke. Beim Flashen probiere ich immer solange bis es funktioniert. Das könnte ich jetzt so nicht sagen, welchen der beiden USB-C Anschlüsse ich da verwende. Da kann ich aber mal drauf achten.

**Reply by DerHavey1994** [reply]
In seiner Mapping hat @Snusme83 43 & 44 für die Battery ausgewählt. UART0 ist also "doppelt belegt". Dadurch verstehe ich, dass bei angeschlossener Battery dann kein flashen möglich ist?!

**Reply by schlimmchen** [reply]
Die Frage ist: Hast du an UART0 TX oder UART0 RX Peripherie angeklemmt, z.B. einen Pin des CAN-Moduls für die Pylontech? Falls ja => schlecht. Dennoch sollte es dann am nativen USB Port immer zuverlässig funktionieren. Aber auch da gibt es Ausnahmen. In jedem Fall sollte "BOOT" die ganze Zeit gedrückt zu halten helfen. Wie es inzwischen in der Doku steht.

**Reply by schlimmchen** [reply]
@DerHavey1994 Danke, du hast gut aufgepasst. Also, @Snusme83, du bist betroffen, und.... lies nochmal die Doku, dann verstehst du, warum du Probleme hast.

**Reply by Snusme83** [reply]
Jetzt glaub ich hab ich es geschnallt. Mit UART0 sind beim ESP32-S3 die GPIO43 und GPIO44 gemeint. Stimmt auf denen hab ich TX (44) und RX (43) von den Pylontechs. D. h. ich muss diese zwei frei kriegen und andere GPIO verwenden und dann spare ich mir auch das Abstecken beim Flashen der Firmware? Das Displayproblem wenn das OLED das 5 Gerät (3 x Smartsolar, 1 x Pylontech, 1x OLED) ist, wird aber damit nicht gelöst, oder?

**Reply by schlimmchen** [reply]
So ist es.

**Reply by Snusme83** [reply]
Alles klar. Ein ESP32-S3 mit Firmware generic_esp32s3_usb kann 3 UARTS bedienen. 

Es ist quasi limitiert auf 3! 

Victron Laderegler, Victron Smartshunts und JK BMS benötigen UARTS. Ich verwende ja 3 x Victron Laderegler - alles aufgebraucht also?

Die Pylontechs benötigen dann keinen UART, das Display hingegen schon oder gehen doch 4 UARTs obwohl in der Limitation von 3 steht und die Pylontechs sind dann Nr. 4 und das Display wäre Nr. 5?

![UARTs](https://github.com/user-attachments/assets/b8d7b33a-8544-47db-9f8e-5fb6f21dec75)

**Reply by schlimmchen** [reply]
3 **hardware** UARTs. Und dann noch ein paar, die in Software mit Interrupts realisiert werden.

> Victron Laderegler, Victron Smartshunts und JK BMS benötigen UARTS. Ich verwende ja 3 x Victron Laderegler - alles aufgebraucht also?

Ja.

Das Display hat kein UART interface, sondern I²C. Wenn es das gleiche Problem ist wie bei @DerHavey1994 dann ist das ein Software-Absturz durch einen dummen Fehler und keine Limitierung in der Hardware. Schau mal bitte auf die serielle Konsole deines Boards, wenn es startet, da ist vermutlich eine Exception und ein Backtrace zu finden.

**Reply by Snusme83** [reply]
![IMG_2196](https://github.com/user-attachments/assets/51b21a9a-5b40-4f72-a7d3-bead7630b029)

Hier soll ich schauen, wenn ich das Profil mit Display in den Hardwareeinstellungen speichere, wenn es abstürzt? Dann muss ich vermutlich mit einem Endgerät (Laptop) die Konsole laufen lassen und mit einem zweiten Endgerät das speichern des Profils übernehmen? Kann ich morgen mal machen, wenn euch das weiterhilft.

**Reply by schlimmchen** [reply]
Nein, der ESP32 läuft ja bei diesem Fehler nicht, dann kannst du auch keine Daten aus der Webkonsole kopieren. Siehe [Wiki](https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Wie-man-nach-L%C3%B6sungen-sucht%2C-Fragen-stellt%2C-Probleme-meldet#42-logdatei-w%C3%A4hrend-des-bootvorgangs-erstellen). Das ist aber nicht zwingend nötig, dass du dich da einfuchst. @DerHavey1994 hat schon einen Backtrace erstellt und ich krieg vermutlich raus, was da schief läuft.

**Comment by Snusme83** [comment]
Soviel wie ihr mir schon geholfen habt, kann ich morgen gerne nochmal einen Absturz provozieren und die serielle Konsole laufen lassen, wenn es der Sache dienlich ist. 🤷🏻‍♂️ Muss morgen eh das pinmapping ändern um RX und TX vom GPIO 43 und 44 wegzubekommen. Dann muss ich hoffentlich nicht mehr zum flashen abstecken.😎

**Reply by Snusme83** [reply]
Servus zusammen,

hab das Speichern des Profiles mit SH1106 soeben nochmal probiert. Wie zu erwarten war, hat sich die OpenDTUonBattery aufgehangen.

Mit putty habe ich über die serielle Konsole folgende Datei erhalten: 

[putty.log](https://github.com/user-attachments/files/16762013/putty.log)

Ich hoffe, das hilft euch weiter.

P.S.: Flashen funktionierte diesmal auch mit angesteckten Pylontechs und die openDTUonBattery läuft jetzt wieder, wie es sein soll.

**Reply by schlimmchen** [reply]
```
0x40377e5e: panic_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/panic.c:408
0x4037d5f5: esp_system_abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/esp_system.c:137
0x403841b9: abort at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/newlib/abort.c:46
0x4037d5eb: _esp_error_check_failed at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/esp_system/esp_err.c:43
0x42077e2d: twai_driver_install at /home/runner/work/esp32-arduino-lib-builder/esp32-arduino-lib-builder/esp-idf/components/driver/twai.c:465 (discriminator 1)
0x42004061: BatteryCanReceiver::init(bool, char const*) at /home/beki/Documents/OpenDTU-OnBattery/src/BatteryCanReceiver.cpp:34
0x4202f7e4: PylontechCanReceiver::init(bool) at /home/beki/Documents/OpenDTU-OnBattery/src/PylontechCanReceiver.cpp:10
0x42003d0d: BatteryClass::updateSettings() at /home/beki/Documents/OpenDTU-OnBattery/src/Battery.cpp:69
0x42003e5b: BatteryClass::init(Scheduler&) at /home/beki/Documents/OpenDTU-OnBattery/src/Battery.cpp:31
0x4204d856: setup() at /home/beki/Documents/OpenDTU-OnBattery/src/main.cpp:197
0x420743b6: loopTask(void*) at /home/beki/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:42
```

Wie sieht das Display denn in diesem Fall aus? Steht da schon "OpenDTU!"?

Wie sieht deine neue pin_mapping.json aus? Hast du Pins 1 oder 2 als battery.rx oder battery.tx verwendet?

Was passiert, wenn du battery.rx und battery.tx testweise auf -1 stellst? Startet die Kiste dann normal?

**Reply by DerHavey1994** [reply]
abort() was called at PC 0x4037d5eb on core 1

Der Abschnitt ist bei mir identisch gewesen.

Display hat bei mir "OpenDTU!" angezeigt & bei uns beiden sind die PINs 1+2 für die LEDs konfiguriert.

**Reply by Snusme83** [reply]
Das Display hab ich noch gar nicht mit dem Mikrocontroller verbunden.

Die funktionierende pinmapping ohne Display ist diese: 
[pin_mapping.json](https://github.com/user-attachments/files/16764113/pin_mapping.json)

Die nicht funktionierende mit Display ist diese:
[pin_mapping.json](https://github.com/user-attachments/files/16764119/pin_mapping.json)

battery.rx und battery.tx soll ich mal testweise im pinmapping auf -1 stellen, anstatt 41 und 42?

**Reply by schlimmchen** [reply]
Ja, bitte. Und testweise mal einen MPPT weniger definieren statt das Display nicht zu definieren. Also immer genau ein Peripheriegerät weniger. Ich hab die dunkle Befürchtung, dass egal welches Gerät aus dem pin mapping rausgeschmissen wird und daher nicht initialisiert wird, zum Erfolg führt (also dass die Kiste dann wieder bootet).

**Reply by Snusme83** [reply]
Also battery.rx und battery.tx auf -1 gesetzt bei der funktionierenden pinmapping. Da passiert gar nix, nur dass es mir in der Live-Ansicht den Akkustand nicht mehr anzeigt, logischerweise.

Jetzt soll ich noch probieren aus dem pinmapping einen Laderegler entfernen (also auf -1 setzen) und anstatt des Ladereglers das Display hinzufügen -  so hätte ich das jetzt verstanden.

**Reply by Snusme83** [reply]
![Pinmapping aktuell Änderung](https://github.com/user-attachments/assets/c21d2913-8e97-4c3f-81fe-6a120d91701f)


So soll ich das pinmapping ändern und dann testen, oder?

**Reply by Snusme83** [reply]
Hab die pinmapping mal verändert wie auf dem oberen Bild zu sehen. Dann funktioniert die OpenDTUonBattery auch, nur, dass halt ein Laderegler weniger angezeigt wird - kein Einfrieren!
[pin_mapping.json](https://github.com/user-attachments/files/16764705/pin_mapping.json)

**Reply by Snusme83** [reply]
Jetzt kommt es mir erst!

auf dem jungfräulichen ESP32-S3 habe ich ja auch die pinmapping mit Display drauf gemacht. Da funktioniert das Display komischerweise, obwohl drei Laderegler, eine Batterie und ein Display im pinmapping hinterlegt sind!🤔
![IMG_2207](https://github.com/user-attachments/assets/20c2974a-9f2a-4e03-8195-231b44b2e18e)

**Reply by Snusme83** [reply]
Komischerweise friert da Nix ein!
![IMG_2208](https://github.com/user-attachments/assets/a396b659-03fb-4888-8473-76930e7cf11a)
![IMG_2209](https://github.com/user-attachments/assets/5f8159f6-762c-4a4a-a468-5fff28286de3)

Das müsste ja dann Hardwareseitig ein Problem sein!🤔

**Reply by DerHavey1994** [reply]
& wenn du jetzt bei den LEDs PINs vergibst? Stürzt er denn wieder ab?

10 & 11 oder so

**Reply by Snusme83** [reply]
Muss ich morgen mal probieren. Ich hab bei allen pinmappings diese LEDs auf -1!

Was sind das eigentlich für LED's?

**Reply by DerHavey1994** [reply]
Man kann extra LEDs anschließen, die die Verbindung zum Inverter & Netzwerk anzeigen.

Beim Board von @swingstate zB sind diese fest verlötet, somit habe ich diese auch ausgewählt.

**Comment by frank-tanne** [comment]
Bei den esp32s3 hatte ich auch das problem mit meinen Display. Am 3,3v Ausgang bricht die Spannung ein ,Kondensator habe ich verbaut. Ich hatte kaum empfang zum Wechselrichter. Jetzt bin direkt mit Display + auf 5V zwei Dioden in reihe dazwischen und gut iss

**Reply by SW-Niko** [reply]
Guter Hinweis,
auf den Demo-Boards ist meistens ein einfacher LDO verbaut, der aus den 5V die 3V3 macht. Auf meinem Board ein SGM2212-3.3XKC3G und der kann laut Datenblatt maximal 800mA. Das kann auch mal eng werden wenn das für den ESP32-S3 + die ganzen externen Komponenten reichen soll.
